날씨 앱 [STEP2] roks, bello

KimRoks opened the pull request:
리뷰어: @protocorn93

버드: @JinUng41, @KimRoks

안녕하세요, 콘!!

STEP 2에서는 위도와 경도를 알아내고, 주소를 변환하는 방법과 알아온 위도, 경도로 현재 날씨 그리고 5일 기상 예보를 받아오는 동작을 구현해 보았습니다.

이번 STEP도 잘 부탁드립니다~ 😄😄
### 요구사항

- [x]  CoreLocation을 통해 현재 좌표 가져오기
- [x]  `CoreLocation`을 통해 주소 가져오기
- [x]  `URLSession`을 통해 API 호출 (HTTP GET 메서드 명시적 호출)
- [x]  전달받은 `JSON` 데이터 파싱 및 모델에 매칭

## 주요 구현 내용

### 디렉터리 구조
```
.
├── WeatherAPI.plist
├── WeatherForecast
│   ├── Controllers
│   │   └── ViewController.swift
│   ├── Extensions
│   │   └── Bundle.swift
│   ├── Info.plist
│   ├── Models
│   │   ├── Locaion
│   │   │   ├── LacationData.swift
│   │   │   └── LocationManager.swift
│   │   ├── Network
│   │   │   ├── NetworkError.swift
│   │   │   └── NetworkManager.swift
│   │   └── Weather
│   │       ├── Current.swift
│   │       ├── Forecast.swift
│   │       ├── WeatherRequest.swift
│   │       └── WeatherType.swift
│   ├── Protocols
│   │   ├── APIRequest.swift
│   │   ├── URLSessionDataTaskProtocol.swift
│   │   └── URLSessionProtocol.swift
│   ├── Resources
│   │   ├── AppDelegate.swift
│   │   ├── Assets.xcassets
│   │   │   ├── AccentColor.colorset
│   │   │   │   └── Contents.json
│   │   │   ├── AppIcon.appiconset
│   │   │   │   └── Contents.json
│   │   │   └── Contents.json
│   │   └── SceneDelegate.swift
│   └── Views
│       └── Base.lproj
│           ├── LaunchScreen.storyboard
│           └── Main.storyboard
```

구현 결과 - JSON Parsing
<img width="1649" alt="스크린샷 2023-11-25 오후 9 21 02" src="https://github.com/tasty-code/ios-weather-forecast/assets/113083860/8e1feddd-10da-4f0e-9559-b2965252935b">




### CoreLocation

- CoreLocation을 이용해 사용자의 위도,경도를 가져오도록 하였습니다.
- 또한 info.plist에 Privacy - Location When In Use Usage Description항목을 추가하여 사용자에게 안내 문구를 표시하도록 하고,
- `locationManagerDidChangeAuthorization()` 내부에서 `requestLocation()`을 사용하여 권한이 바뀔때 위치정보를 업데이트 하도록 하였습니다
    - `startUpdatingLocation()`으로 호출하게 되면 `stopUpdatingLocation()`이 호출될 때까지 계속 새로운 위치 정보를 가져오기 때문에,
    - 한번만 동작하는 `requestLocation()`을 사용하게 되었습니다.
   
<img width="163" alt="Untitled" src="https://github.com/tasty-code/ios-weather-forecast/assets/113083860/98783f93-9626-4751-ad10-b9863cabd813">

### CLGeocoder & CLPlacemark

날씨 API에서 제공하는 기본적인 지역 정보보다 사용자에게 조금 더 익숙하고 정확한 위치 정보를 제공하기 위해

CLGeocoder & CLPlacemark를 채택하였습니다.

Placemark에서 7개의 위치를 제공하였는데

국내에서는 `locality` 등의 경우 비수도권에선 표시되지 않아

아래 3개를 사용하여 수도권과 비수도권 모두 같은 수준의 위치정보를 제공 받도록 하였습니다

1. administrativeArea
    - 플레이스마크와 관련된 주 또는 도입니다.
    - ex) 서울특별시, 강원도
2. subAdministrativeArea
    - 플레이스마크에 대한 추가 행정 구역 정보입니다.
    - ex) 횡성군
3. subLocalty
    - 플레이스마크에 대한 추가 도시 수준 정보입니다.
    - ex) 도포면,법원읍

## 궁금한 점

### 프로토콜의 사용

- 현재 두개의 api 통신 작업을 하나의 정의된 프로토콜을 채택하여 진행하고 있습니다
하지만 데이터를 받아오는 과정에선 결국 request객체를 2번 만들게 되는데 이러한 방식이 옳은지 의문이 듭니다

이러한 방식은 protocol의 좋은 사용 예시일까요? 보충을 한다면 어떻게 해야할까요? 의견 부탁드립니다.

### 사용자의 위치 데이터를 얻지 못했을때의 처리

- CoreLoaction을 학습하다가 앱이 어떠한 이유로 위치 데이터를 사용할 수 없을 때도 앱 사용 경험을 최적화 하라고 명시 되어 있었고 미리 기능을 비활성화 하는것이 도중에 오류에 대응하는것보다 더 안정적인 접근방식이라고 하던데, 

날씨앱 같은 경우 위치데이터에 대한 의존도가 굉장히 높기 때문에 이런 경우 어떻게 해야할지 고민입니다.

1. 위치 데이터를 허용하지 않으면 앱을 이용할 수 없다는 안내를 한 후 앱을 종료한다.
2. 특정 지역의 날씨를 임의적으로 제공한다.

위 두개와 같은 생각을 해봤는데 더 좋은 방식이 있다면 어떤 방식이 있을까요? 콘의 의견이 궁금합니다.

JinUng41 committed:
feat: info.plist에 위치 접근 허용을 위한 Privacy 정의

- Privacy Location When In Usage Description

JinUng41 committed:
feat: 폴더링 및 필요 파일 추가 생성

- 위도, 경도, 주소를 저장하는 LocationData
- 받아올 기상 예보에 대한 두 가지를 정의하는 enum WeatherType
- 실제 위치 정보를 받고 주소를 변환하는 LocationManager

JinUng41 committed:
refactor: 변수명 수정 및 상수 선언

- 변하지 않아야 하는 내용에 대해서 let으로 선언
- error가 발생했을 때, `return`을 통해 early exit

JinUng41 committed:
feat: API 요청을 만들고 네트워크 작업을 실제 수행

- 프로토콜을 채택하고 실제 동작을 정의
- API 요청을 만들고, 실제 네트워크 작업을 수행

JinUng41 committed:
refactor: 주소 변환을 placemarks의 마지막 요소로 함.

- 불필요한 변수 삭제
- 변수 네이밍 수정

JinUng41 committed:
test: 기능 확인을 위한 코드 삽입

KimRoks committed:
refactor: 파일 이름 변경

- LocationInfo -> LocationData

protocorn93 commented on WeatherForecast/WeatherForecast/Controllers/ViewController.swift:
```diff
@@ -5,10 +5,50 @@
+    }
+}
+
+extension ViewController {
+    private func printLocationData(_ data: LocationData) {
+        print("위도: \(data.latitude) 경도: \(data.longitude) 주소: \(data.address) \n")
+    }
+}
```
디버깅을 위한 코드로 보여지는데요. 요런 코드는 제거하고 올려주시는게 좋아요 ㅎㅎ

JinUng41 replied:
네, 다음 스텝부터는 제거하고 올리도록 하겠습니다!!

protocorn93 commented on WeatherForecast/WeatherForecast/Controllers/ViewController.swift:
```diff
@@ -5,10 +5,50 @@
+        networkManager.fetchData(for: weatherRequest) { (result: Result<Current, Error>) in
+            switch result {
+            case .success(let weather):
+                print(weather)
+                print("")
```
마찬가지로 디버깅을 위한 코드는 제거하고 PR 올려주시면 좋을 것 같아요. 다음 스탭에서 구현될 것이기 때문에, `break` 정도만 추가해두면 좋겠네요.

JinUng41 replied:
네, 감사합니다!! 👍👍

protocorn93 commented on WeatherForecast/WeatherForecast/Controllers/ViewController.swift:
```diff
@@ -5,10 +5,50 @@
+import CoreLocation
 
 final class ViewController: UIViewController {
+    private let locationManager: LocationManager = LocationManager()
+    private let networkManager: NetworkManager = NetworkManager()
```
ViewController 생성 시점에 주입해줄 수 있는 방법을 고민해보셔도 좋을 것 같아요.

KimRoks replied:
Scene Delegate를 이용해 ViewContoller의 생성 시점에 두 객체의 의존성을 주입하도록 변경해보았습니다
이에 따라 StoryBoard는 삭제 조치하였습니다

protocorn93 commented on WeatherForecast/WeatherForecast/Models/Locaion/LocationManager.swift:
```diff
@@ -0,0 +1,80 @@
+import CoreLocation
+
+protocol LocationManagerDelegate: AnyObject {
+    func update(with data: LocationData)
```
사소하지만 위임을 담당하는 객체이기 때문에 `didUpdateLocation`과 같이 변화, 상태를 공유해줄 수 있는 인터페이스명으로 작성해주시는 것이 좋을 것 같아요.

JinUng41 replied:
delegate를 지정하기 위한 프로토콜이기 때문에 `UpdatedLocationDelegate`라고 바꾸어 보았습니다. 어떻게 생각하실까요?

KimRoks replied:
리팩터링 작업 중 해당 프로토콜의 기능이 수정되었습니다
그 과정에서 프로토콜의 네이밍을 `LocationUpdateDelegate`로 변경하였는데
기본적으로 청사진의 역할을 하는 프로토콜의 경우 ~ able로 네이밍하고 Delegate 관련 프로토콜은 Delegate라는 네이밍을 하기 원했기때문에
`명확한 기능 + Delegate`로 변경해보았습니다
확인해보시고 콘의 의견도 들려주세요!

protocorn93 commented on WeatherForecast/WeatherForecast/Models/Locaion/LocationManager.swift:
```diff
@@ -0,0 +1,80 @@
+    override init() {
+        super.init()
+        locationManager.delegate = self
+        locationManager.desiredAccuracy = kCLLocationAccuracyKilometer
```
`desiredAccuracy`를 `kCLLocationAccuracyKilometer`로 할당해준 이유가 있을까요?

KimRoks replied:
학습한 내용을 토대로 실험을 해보았습니다.
높은 정확도를 요구할수록 데이터를 가져오는 속도가 느려지고 전력효율면에서도 좋지 않아 사용자의 입장에서 좋지 않다 판단하였습니다.
선택지는 크게 100m, 1km, 3km 였는데
날씨앱 수준에서 크게 어긋나지 않는 주소를 제공하고 속도면에서도 `kCLLocationAccuracyKilometer`이 적합하다고 판단하여 선택하게 되었습니다.

protocorn93 replied:
선택한 의도가 좋네요👍

protocorn93 commented on WeatherForecast/WeatherForecast/Models/Locaion/LocationManager.swift:
```diff
@@ -0,0 +1,80 @@
+        let geocoder = CLGeocoder()
+        geocoder.reverseGeocodeLocation(location) { [weak self] placemarks, error in
+            if let error = error {
+                print("주소 가져오기 실패: \(error.localizedDescription)")
```
에러를 사용자에게 친절하게 알려줄 필요도 있어요. 에러를 외부로 전달해줄 수 있으면 좋겠어요. 그리고 화면에선 알럿을 통해 보여주면 더 좋을 것 같구요.

JinUng41 replied:
프로토콜에 메서드를 추가로 선언하여 에러가 발생했을 때는 Alert를 띄울 수 있도록 하였습니다.
허나 error를 직접 인자로 전달하지 않도록 수정한 이유는 사용자가 확인하는 Alert에서 error에 대해 구체적으로 알 필요가 없다고 판단하였습니다.
따라서 error를 인자로 전달하지 않고 단순히 특정 Alert를 띄울 수 있도록 하였습니다.

protocorn93 commented on WeatherForecast/WeatherForecast/Models/Locaion/LocationManager.swift:
```diff
@@ -0,0 +1,80 @@
+    }
+    
+    func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
+        print("위치 업데이트 실패: \(error.localizedDescription)")
```
마찬가지로 에러를 외부로 전달할 수 있으면 좋겠어요.

JinUng41 replied:
네 수정하였습니다.

protocorn93 reviewed (changes requested):
두 분 모두 고생하셨어요. 깔끔하게 작성해주신 코드와 커밋로그 덕분에 여러분들의 작성 의도를 보다 잘 파악할 수 있었어요! 

`CLLocationManagerDelegate`를 숨기고 `LocationManagerDelegate`로 외부와 소통한 점이 인상적이었어요.


## 질문

프로토콜의 사용

> 현재 두개의 api 통신 작업을 하나의 정의된 프로토콜을 채택하여 진행하고 있습니다
하지만 데이터를 받아오는 과정에선 결국 request객체를 2번 만들게 되는데 이러한 방식이 옳은지 의문이 듭니다
이러한 방식은 protocol의 좋은 사용 예시일까요? 보충을 한다면 어떻게 해야할까요? 의견 부탁드립니다.

이건 프로토콜의 사용 예시와 별개로 서로 다른 두 api 호출(현재 날짜, 최근 날씨 예보)이기 때문에 두 번의 요청을 하는 것이 맞아요. 

프로토콜로 네트워크 객체를 추상화 해주시는 것은 옳은 방향이었어요 👍

> 사용자의 위치 데이터를 얻지 못했을때의 처리
>
> CoreLoaction을 학습하다가 앱이 어떠한 이유로 위치 데이터를 사용할 수 없을 때도 앱 사용 경험을 최적화 하라고 명시 되어 있었고 미리 기능을 비활성화 하는것이 도중에 오류에 대응하는것보다 더 안정적인 접근방식이라고 하던데,
날씨앱 같은 경우 위치데이터에 대한 의존도가 굉장히 높기 때문에 이런 경우 어떻게 해야할지 고민입니다.
>
> 위치 데이터를 허용하지 않으면 앱을 이용할 수 없다는 안내를 한 후 앱을 종료한다.
특정 지역의 날씨를 임의적으로 제공한다.
위 두개와 같은 생각을 해봤는데 더 좋은 방식이 있다면 어떤 방식이 있을까요? 콘의 의견이 궁금합니다.

이건 무엇이 옳다 그르다는 아닌 것 같아요. 서비스가 지향하는 방향에 따라 달라지겠죠. 

오로지 위치기반의 날씨앱이라면 위치 허용이 되지 않으면 서비스를 이용할 수 없다고 사용자에게 알려야겠죠. 

위치정보는 민감한 개인정보이기 때문에 누군가는 이러한 행위에 대해 불친절하고 불쾌하다고 생각할 수 있어요. 

그것이 아니랴면 개인적으론 도시 검색을 통해 해당 도시의 날짜를 보여주는 기능을 추가할 것 같아요. 위치 정보를 제공하지 않는 사용자들도 사용할 수 있도록 말이죠! (애플의 날씨앱 처럼)  여기까지는 서비스에 관한 저의 생각이었구요. 

언제든 사용자가 원한다면 위치 권한을 바꿀 수 있는 방법을 제시해주거나 위치 권한 팝업을 띄울 수 있는 구조에 대해서도 고민해보시면 좋을 것 같아요. .denied, .notDetermied 등의 케이스에 대응도 이에 들어갈 수 있겠죠. 

아이폰 설정 앱의 여러분 앱의 위치 설정 화면으로 이동 시켜주는 방법도 있을 것이구요.

KimRoks committed:
refactor: 피드백 반영

- delegate 프로토콜의 네이밍 변경
- 디버깅 코드의 삭제 조치 및 수정

KimRoks committed:
refactor: 피드백적용

- 에러를 ViewController로 전달하기 위해 Delegate프로토콜의 수정 및 의도에 맞게끔 리네임
- placemark를 가져오지 못하는 경우는 아주 예외적이므로 주소 주소 자체를 가져오지 못할 경우 early exit 하도록 메서드 수정
- 에러 상황시 Alert를 표시하도록 추가

JinUng41 committed:
refactor: 불필요한 파일 삭제

JinUng41 committed:
refactor: Alert를 띄우는 프로토콜 메서드에서 error를 인자로 받지 않음.

- 사용자 입장에서 error의 내용을 알 필요가 없다고 판단하였음.
- 따라서 error를 인자로 받지는 않게 하였음.

KimRoks committed:
refactor: Di 수정

- SceneDelegate를 이용하여 locationManager,NetworkManager 두 객체의 의존성을 ViewController 생성시에 주입할 수 있도록 변경
- StoryBoard 삭제
